### PR TITLE
Add Consul Catalog provider 

### LINKED ISSUE #4675: Add a provider for ConsulCatalog in Traefik v2
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Traefik v1 supports dynamic config generation for and from a Consul Catalog service, i.e. it has a ConsulCatalog provider.

Now that the main changes for v2 are out, we can re-add previously supported providers, such as the one for ConsulCatalog.

The existing providers, as well as the code for the ConsulCatalog provider in v1.7, can be used for inspiration.

### PR DESCRIPTION
**Pull request by negasus**
### What does this PR do?

Add new provider 'Consul Catalog'

### Motivation

Popular feature of Traefik v1
Close #4675

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### CODE REVIEW

**Comment by negasus** [comment]
Added a couple integration tests.
But docker integration test fails on fresh repo. So in my PR also.

FAIL: docker_test.go:147: DockerSuite.TestDockerContainersWithTCPLabels

File: pkg/provider/consulcatalog/build_configuration_test.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
+package consulcatalog
+
+import (
+	"context"
+	"github.com/containous/traefik/v2/pkg/config/dynamic"
+	"github.com/stretchr/testify/require"
+	"gotest.tools/assert"
```
Comment (ldez) [inline pkg/provider/consulcatalog/build_configuration_test.go]:
```suggestion
		"github.com/stretchr/testify/assert"
```

File: go.mod [hunk 2]
```diff
@@ -98,7 +99,7 @@ require (
 	gopkg.in/fsnotify.v1 v1.4.7
 	gopkg.in/jcmturner/goidentity.v3 v3.0.0 // indirect
 	gopkg.in/yaml.v2 v2.2.2
-	gotest.tools v2.2.0+incompatible // indirect
+	gotest.tools v2.2.0+incompatible
```
Comment (ldez) [inline go.mod]:
```suggestion
	gotest.tools v2.2.0+incompatible // indirect
```

**Comment by ldez** [comment]
Could you rebase (not merge) on the `master` branch.

And also add integration tests :wink:

**Comment by ldez** [comment]
could you clean your commits.

**Comment by negasus** [comment]
Yea, sorry) I try clean

**Comment by mmatur** [comment]
add `no-merge` finalizing the documentation

**Comment by megalan247** [comment]
Would it be possible to add a custom configuration file option similar to traefik v1? We have a heavily customized configuration and it would be really nice if we could specify the config directly, instead of relying on "defaultRule" which can be quite limiting.

For example, we have two services named the same, but one tagged with stage and one with prod, and we need logic to check if a certain tag exists and assign it to a different backend.

**Comment by imyller** [comment]
Looks like this is ready to merge and only pending @ldez review?

I think many Traefik users are eagerly waiting for this feature.

Thank you @negasus, @mmatur and @ldez for your efforts so far 👍

**Comment by ldez** [comment]
https://docs.traefik.io/contributing/maintainers/#pr-review-process

**Comment by frederikbosch** [comment]
The last comment means, I guess, we need at least three maintainer approvals before merging. Are we it that of the process already @ldez ?

File: pkg/provider/consulcatalog/config_test.go [hunk 3]
```diff
@@ -0,0 +1,1849 @@
...
+					Middlewares: map[string]*dynamic.Middleware{},
+					Services: map[string]*dynamic.Service{
+						"Test": {
+							LoadBalancer: &dynamic.ServersLoadBalancer{
+								Servers: []dynamic.Server{
+									{
+										URL: "http://127.0.0.1:80",
+									},
+								},
+								PassHostHeader: true,
```
Comment (sodabrew) [inline pkg/provider/consulcatalog/config_test.go]: I think this needs to be `PassHostHeader: Bool(true)`

**Comment by sodabrew** [comment]
🍾

**Pull request merged**